File: lib/sciolyff/interpreter/tournament.rb
# frozen_string_literal: true

require 'sciolyff/interpreter/model'
require 'sciolyff/interpreter/bids'

module SciolyFF
  # Models a Science Olympiad tournament
  class Interpreter::Tournament < Interpreter::Model
    def initialize(rep)
      @rep = rep[:Tournament]
    end

    def link_to_other_models(interpreter)
      @events = interpreter.events
      @teams = interpreter.teams
      @placings = interpreter.placings
      @penalties = interpreter.penalties
      @tracks = interpreter.tracks
    end

    attr_reader :events, :teams, :placings, :penalties, :tracks

    undef tournament

    %i[
      name
      location
      level
      state
      division
      year
    ].each do |sym|
      define_method(sym) { @rep[sym] }
    end

    def short_name
      @rep[:'short name']
    end

    def date
      @date ||= if @rep[:date].instance_of?(Date)
                  @rep[:date]
                else
                  unless @rep[:date].nil?
                    Date.parse(@rep[:date])
                  end
                end
    end

    def start_date
      @start_date ||= if !@rep[:'start date'].nil?
                        if @rep[:'start date'].instance_of?(Date)
                          @rep[:'start date']
                        else
                          Date.parse(@rep[:'start date'])
                        end
                      else
                        date
                      end
    end

    def end_date
      @end_date ||= if !@rep[:'end date'].nil?
                      if @rep[:'end date'].instance_of?(Date)
                        @rep[:'end date']
                      else
                        Date.parse(@rep[:'end date'])
                      end
                    else
                      start_date
                    end
    end

    def awards_date
      @awards_date ||= if !@rep[:'awards date'].nil?
                         if @rep[:'awards date'].instance_of?(Date)
                           @rep[:'awards date']
                         else
                           Date.parse(@rep[:'awards date'])
                         end
                       else
                         end_date
                       end
    end

    def medals
      @rep[:medals] || [calc_medals, maximum_place].min
    end

    def trophies
      @rep[:trophies] || [calc_trophies, nonexhibition_teams_count].min
    end

    def bids
      @rep[:bids] || 0
    end

    def bids_per_school
      @rep[:'bids per school'] || 1
    end

    def worst_placings_dropped?
      worst_placings_dropped.positive?
    end

    def worst_placings_dropped
      @rep[:'worst placings dropped'] || 0
    end

    def exempt_placings?
      exempt_placings.positive?
    end

    def exempt_placings
      @rep[:'exempt placings'] || 0
    end

    def custom_maximum_place?
      maximum_place != nonexhibition_teams_count
    end

    def maximum_place
      @rep[:'maximum place'] || nonexhibition_teams_count
    end

    def per_event_n
      @rep[:'per-event n']
    end

    def per_event_n?
      @rep.key? :'per-event n'
    end

    def n_offset
      @rep[:'n offset'] || 0
    end

    def ties?
      @ties ||= placings.map(&:tie?).any?
    end

    def ties_outside_of_maximum_places?
      @ties_outside_of_maximum_places ||= placings.any? do |p|
        p.tie? && !p.points_limited_by_maximum_place?
      end
    end

    def tracks?
      !@tracks.empty?
    end

    def nonexhibition_teams_count
      @nonexhibition_teams_count ||= @teams.count { |t| !t.exhibition? }
    end

    include Interpreter::Bids

    private

    def calc_medals
      (nonexhibition_teams_count / 10r).ceil.clamp(3, 10)
    end

    def calc_trophies
      (nonexhibition_teams_count / 6r).ceil.clamp(3, 10)
    end
  end
end

File: src/interpreter/tournament.ts
import Model from "./model.js";
import type {
  TournamentRep,
  Interpreter,
  Event,
  Team,
  Placing,
  Track,
  Penalty,
} from "./types.js";

export default class Tournament implements Model<TournamentRep> {
  rep: TournamentRep;

  // references (access after link)
  events?: Event[];
  teams?: Team[];
  placings?: Placing[];
  tracks?: Track[];
  penalties?: Penalty[];

  // rep properties
  location: string;
  level: string;
  division: string;
  year: number;
  name?: string;
  state?: string;
  medals?: number;
  trophies?: number;
  bids?: number;
  bidsPerSchool?: number;
  shortName?: string;
  worstPlacingsDropped?: number;
  exemptPlacings?: number;
  reverseScoring: boolean;
  maximumPlace?: number;
  perEventN?: string;
  nOffset?: number;
  date?: Date;
  startDate?: Date;
  endDate?: Date;
  awardsDate?: Date;
  testRelease?: string;

  // assorted calculated things
  // not sure if these are necessary?
  // just here because they're defined in the ruby version
  hasCustomMaximumPlace?: boolean;
  hasTies?: boolean;
  hasTiesOutsideOfMaximumPlaces?: boolean;
  hasTracks?: boolean;
  // for reverse scoring
  largestPlace?: number;

  nonExhibitionTeamsCount?: number;

  _topTeamsPerSchool?: Team[];
  _teamsEligibleForBids?: Team[];

  constructor(rep: TournamentRep) {
    this.rep = rep;

    this.location = rep.location;
    this.level = rep.level;
    this.division = rep.division;
    this.year = rep.year;
    this.name = rep.name;
    this.state = rep.state;

    this.medals = rep.medals;
    this.trophies = rep.trophies;

    this.bids = rep.bids ?? 0;
    this.bidsPerSchool = rep["bids per school"] ?? 1;
    this.shortName = rep["short name"];
    this.worstPlacingsDropped = rep["worst placings dropped"] ?? 0;
    this.exemptPlacings = rep["exempt placings"] ?? 0;
    this.reverseScoring = rep["reverse scoring"] ?? false;

    this.maximumPlace = rep["maximum place"];

    this.perEventN = rep["per-event n"];
    this.nOffset = rep["n offset"] ?? 0;

    this.date = rep.date ? new Date(rep.date) : undefined;
    this.startDate = rep["start date"]
      ? new Date(rep["start date"])
      : this.date;
    this.endDate = rep["end date"] ? new Date(rep["end date"]) : this.startDate;
    this.awardsDate = rep["awards date"]
      ? new Date(rep["awards date"])
      : this.endDate;
    this.testRelease = rep["test release"];
  }

  public link(interpreter: Interpreter): void {
    this.events = interpreter.events;
    this.teams = interpreter.teams;
    this.placings = interpreter.placings;
    this.tracks = interpreter.tracks;
    this.penalties = interpreter.penalties;

    this.nonExhibitionTeamsCount = this.teams.filter(
      (t) => !t.exhibition
    ).length;
    this.maximumPlace = this.maximumPlace ?? this.nonExhibitionTeamsCount;

    this.medals = this.medals ?? Math.min(this.calcMedals(), this.maximumPlace);
    this.trophies =
      this.trophies ??
      Math.min(this.calcTrophies(), this.nonExhibitionTeamsCount);

    this.hasCustomMaximumPlace =
      this.maximumPlace !== this.nonExhibitionTeamsCount;
  }

  computeProperties(): void {
    if (!this.placings || !this.tracks) {
      throw new Error("things are undefined");
    }

    this.hasTies = this.placings.some((p) => p.tie);
    this.hasTiesOutsideOfMaximumPlaces = this.placings.some(
      (p) => p.tie && !p.pointsLimitedByMaximumPlace
    );
    this.hasTracks = this.tracks.length > 0;
  }

  computeLargestPlace(): void {
    if (!this.placings) {
      throw new Error("things are undefined");
    }

    if (this.reverseScoring) {
      this.largestPlace = Math.max(...this.placings.map((p) => p.place || 0));
    }
  }

  private calcMedals() {
    // divides by 10, clamps number between 3 and 10
    return Math.min(
      Math.max(Math.ceil((this.nonExhibitionTeamsCount ?? 0) / 10), 3),
      10
    );
  }
  private calcTrophies() {
    // divides by 6, clamps number between 3 and 10
    return Math.min(
      Math.max(Math.ceil((this.nonExhibitionTeamsCount ?? 0) / 6), 3),
      10
    );
  }

  // bids logic
  public get topTeamsPerSchool(): Team[] | undefined {
    // select the first team from each school
    return (this._topTeamsPerSchool ||= Array.from(
      this.teams
        ?.reduce(
          (acc, t) =>
            acc.has(`${t.school}|${t.city ?? ""}|${t.state}`)
              ? acc
              : acc.set(`${t.school}|${t.city ?? ""}|${t.state}`, t),
          new Map<string, Team>()
        )
        .values() ?? []
    ));
  }

  public get teamsEligibleForBids(): Team[] {
    return this.bidsPerSchool === 1
      ? (this.topTeamsPerSchool as Team[])
      : (this._teamsEligibleForBids ||= Array.from(
          this.teams
            ?.reduce((acc, t) => {
              const k = [t.school, t.city, t.state];
              if (acc.has(k)) {
                (acc.get(k) as Team[]).push(t);
              } else {
                acc.set([t.school, t.city, t.state], [t]);
              }
              return acc;
            }, new Map<(string | undefined)[], Team[]>())
            ?.values() as IterableIterator<Team[]>,
          (teams) =>
            teams
              .sort((a, b) => (a.rank as number) - (b.rank as number))
              .slice(0, this.bidsPerSchool)
        )
          .flat()
          .sort((a, b) => (a.rank as number) - (b.rank as number)));
  }
}
